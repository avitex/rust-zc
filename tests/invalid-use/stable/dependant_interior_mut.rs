use std::cell::RefCell;

use zc::Dependant;

#[derive(PartialEq, Debug, Dependant)]
pub struct StructWithBytes<'a>(RefCell<&'a [u8]>);

impl<'a> From<&'a [u8]> for StructWithBytes<'a> {
    fn from(bytes: &'a [u8]) -> Self {
        Self(RefCell::new(&bytes[1..]))
    }
}

fn main() {
    let owner = vec![1, 2, 3];
    let owner2 = vec![6, 6, 6];
    let data = zc::from!(owner, StructWithBytes, [u8]);
    dbg!(data.dependant::<StructWithBytes>());
    *data.dependant::<StructWithBytes>().0.borrow_mut() = &owner2;
    drop(owner2);
    dbg!(data.dependant::<StructWithBytes>());
}
