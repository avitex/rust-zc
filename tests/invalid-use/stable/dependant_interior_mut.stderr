error[E0277]: `RefCell<&'static [u8]>` cannot be shared between threads safely
 --> $DIR/dependant_interior_mut.rs:5:28
  |
5 | #[derive(PartialEq, Debug, Dependant)]
  |                            ^^^^^^^^^ `RefCell<&'static [u8]>` cannot be shared between threads safely
  |
  = help: within `StructWithBytes<'static>`, the trait `Sync` is not implemented for `RefCell<&'static [u8]>`
  = note: required because it appears within the type `StructWithBytes<'static>`
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
